[receiver/promtailreceiver] Add promtail receiver

mar4uk opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Added new promtail receiver component

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9800

**Documentation:** <Describe the documentation added.>

Currently, this receiver supports scrape_config static_configs and scrape_configs loki_push_api. I didn't check other scrape_config targets (syslog, kubernetes, window_events), some additional fixes are probably needed to make them work
The config file will look like this:
```
receivers:
  promtail:
   config:
    positions:
      filename: /tmp/positions.yaml
      sync_period: 10s
    scrape_configs:
      - job_name: system
        static_configs:
          - labels:
              job: varlogs
              __path__: /var/log/example.log
      - job_name: push
        loki_push_api:
          server:
            http_listen_port: 3101
            grpc_listen_port: 3600
          labels:
            pushserver: push1
          use_incoming_timestamp: true
    target_config:
      sync_period: 10s

processors:
  batch:

exporters:
  logging:
    logLevel: debug
  loki:
    endpoint: "http://localhost:3100/loki/api/v1/push"
    labels:
      attributes:
        pushserver:
        foo:

service:
  pipelines:
    logs:
      receivers: [ promtail ]
      processors: [ batch ]
      exporters: [ logging, loki ]
```

To send logs to loki via collector we can run curl:
```
curl -i -XPOST -H "Content-Type: application/json" "http://localhost:3101/loki/api/v1/push" -d '{
  "streams": [
    {
      "stream": {
        "foo": "bar"
      },
      "values": [
          [ "1652782740000000000", "def" ]
      ]
    }
  ]
}'
```
To send logs to loki via collector we tail file example.log. It could be tested with command:
```
echo "May 04 11:45:56 Irinas-MBP login[87256]: USER_PROCESS: 87256 ttys012" >> example.log
```

The initial draft [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/10109) was closed because of uncertainty with the license. According to https://github.com/grafana/loki/blob/main/LICENSING.md#apache-20 `clients/` is under Apache-2.0 license and we can use its code in collector-contrib

**Important**
PR https://github.com/grafana/loki/pull/7320 should be merged first, it contains mapstructure support in promtail

mar4uk committed:
[receivers/promtail] Add promtail input operator to stanza

mar4uk committed:
[receivers/promtail] Add new promtailreceiver component

mar4uk committed:
[receivers/promtail] Add promtailreceiver component to components

djaglowski commented on pkg/stanza/go.mod:
```diff
@@ -13,45 +13,242 @@ require (
 )
 
 require (
+	github.com/go-kit/log v0.2.1
+	github.com/grafana/loki v1.6.2-0.20220927161239-f3b72c8b591e
 	github.com/influxdata/go-syslog/v3 v3.0.1-0.20210608084020-ac565dc76ba6
 	github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage v0.61.0
+	github.com/prometheus/client_golang v1.13.0
+	github.com/prometheus/common v0.37.0
+	github.com/prometheus/prometheus v1.8.2-0.20211119115433-692a54649ed7
```
In the long run, I'd like to reduce the scope of the `pkg/stanza` codebase by moving receiver-specific input operators into the appropriate receiver modules. With that in mind, can I ask that all the promtail related code reside in the new receiver codebase? You should be able to implement and register an operator with exported functions.

mar4uk replied:
moved promtail related code to receiver/promtailreceiver directory

djaglowski commented on pkg/stanza/operator/input/promtail/promtail.go:
```diff
@@ -0,0 +1,165 @@
+	a.client.Stop()
+}
+
+func (operator *PromtailInput) Start(_ operator.Persister) error {
```
`operator` is overloaded here, referring to both the package and this instance of the struct.

mar4uk replied:
thanks! fixed

djaglowski commented on receiver/promtailreceiver/Makefile:
```diff
@@ -0,0 +1 @@
+include ../../Makefile.Common
```
nit
```suggestion
include ../../Makefile.Common

```

djaglowski commented on receiver/promtailreceiver/testdata/config.yaml:
```diff
@@ -0,0 +1,11 @@
+            job: varlogs
+            __path__: testdata/simple.log
+  target_config:
+    sync_period: 10s
```
```suggestion
    sync_period: 10s

```

djaglowski commented on pkg/stanza/operator/input/promtail/promtail_test.go:
```diff
@@ -0,0 +1,173 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/entry"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/testutil"
+)
+
```
You may want to use the `operatortest` package to validate unmarshaling of the config. The nice thing about this package is that it hooks into the collector's actual method of unmarshaling, so you are validating against all of its custom decode hooks. I found this very useful when adding mapstructure support to other operators.

[Here's an example](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/71dd19d2e59cd1f8aa9844461089d5c17efaa0ca/pkg/stanza/operator/parser/syslog/config_test.go#L28).

djaglowski reviewed (commented):
This generally looks good. I have one big ask regarding where the code resides, and otherwise just a few minor things.

mar4uk committed:
[receivers/promtail] update loki, address review comments

mar4uk committed:
[receivers/promtail] add more tests

mar4uk committed:
[receivers/promtail] add code comments

mar4uk committed:
[receivers/promtail] add README.md

djaglowski commented on receiver/promtailreceiver/internal/logger_test.go:
```diff
@@ -0,0 +1,301 @@
+	"go.uber.org/zap/zaptest/observer"
+)
+
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
Please move this to the top of the file.

mar4uk replied:
done

djaglowski commented on receiver/promtailreceiver/promtail.go:
```diff
@@ -0,0 +1,127 @@
+// Copyright 2021 OpenTelemetry Authors
```
```suggestion
// Copyright The OpenTelemetry Authors
```

mar4uk replied:
done

djaglowski commented on receiver/promtailreceiver/factory.go:
```diff
@@ -0,0 +1,48 @@
+package promtailreceiver
```
Please add the license header to all go files.

mar4uk replied:
done

djaglowski commented on go.mod:
```diff
@@ -1011,3 +1044,14 @@ exclude github.com/docker/distribution v2.8.0+incompatible
 // at the time of this inclusion, v0.38.0 was the latest version available (also tagged as v2.38.0)
 replace github.com/prometheus/prometheus => github.com/prometheus/prometheus v0.38.0
+
+replace github.com/hpcloud/tail => github.com/grafana/tail v0.0.0-20220426200921-98e8eb28ea4c
+
+// using fork for support LogpullReceived and LogpullFields for loki - see https://github.com/cloudflare/cloudflare-go/pull/743
+replace github.com/cloudflare/cloudflare-go => github.com/cyriltovena/cloudflare-go v0.27.1-0.20211118103540-ff77400bcb93
+
+// using fork for custom dialer support - see https://github.com/bradfitz/gomemcache/pull/86
+replace github.com/bradfitz/gomemcache => github.com/grafana/gomemcache v0.0.0-20220812141859-1e3ae89e91a7
+
+// Fork containing a line-buffered logger which should improve logging performance.
+replace github.com/go-kit/log => github.com/dannykopping/go-kit-log v0.2.2-0.20221002180827-5591c1641b6b
```
It seems odd that one component would require us to almost double the number of special cases in this file. Is there a general reason that this would necessarily be the case? Is it possible to remove any of these?

mar4uk replied:
That's a good observation. I removed the last replace (github.com/go-kit/log) by downgrading loki a little bit. I can't remove other `replace`s right now because loki needs them. They will be eventually removed when they will be removed from loki

mar4uk committed:
[receivers/promtail] added changes to CHANGELOG.md, added license to promtail receiver go files, removed one replace directive from go mod

mar4uk committed:
[receivers/promtail] fix CHANGELOG.md

mar4uk committed:
[receivers/promtail] fix PR checks

mar4uk commented:
looking into failed checks..

mar4uk commented:
@djaglowski could you please take another look?

djaglowski commented on .github/workflows/build-and-test.yml:
```diff
@@ -75,6 +75,9 @@ jobs:
     runs-on: ubuntu-latest
     needs: [setup-environment]
     steps:
+      - run: sudo apt-get -qq update
+      - name: Install libsystemd-dev
+        run: sudo apt-get install libsystemd-dev
```
For my benefit, will you please explain why this is needed?

mar4uk replied:
On Linux, Promtail requires the systemd headers to be installed for Journal support.
Promtail brought package [sdjournal](https://github.com/coreos/go-systemd/blob/main/sdjournal/journal.go) that uses C.
To make compilation work we need to install libsystemd-dev

djaglowski replied:
Thanks for explaining

djaglowski commented on go.mod:
```diff
@@ -1015,3 +1048,11 @@ exclude github.com/docker/distribution v2.8.0+incompatible
 // at the time of this inclusion, v0.38.0 was the latest version available (also tagged as v2.38.0)
 replace github.com/prometheus/prometheus => github.com/prometheus/prometheus v0.38.0
+
+replace github.com/hpcloud/tail => github.com/grafana/tail v0.0.0-20220426200921-98e8eb28ea4c
```
Please include a comment explaining why this is needed.

mar4uk replied:
done

djaglowski commented on receiver/promtailreceiver/config.go:
```diff
@@ -0,0 +1,117 @@
+
+type Config struct {
+	helper.InputConfig `mapstructure:",squash" yaml:",inline"`
+	// Promtail input config is declared as nested key to allow custom unmarshalling
+	// confmap doesn't call custom unmarshal for the top structure
+	// see the implementation of unmarshalerHookFunc in https://github.com/open-telemetry/opentelemetry-collector:
+	// https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/confmap.go#L274
```
I'm not sure this is accurate. Many other components implement custom unmarshal functions for their top-level configs. I believe the top-level issue only applies when calling unmarshal directly on the struct, instead of allowing the hook func to do so.

In any case, if this ends up being necessary, please use a permalink to the code. It appears to already be slightly out of date as some changes were made to the file.

mar4uk replied:
Thank you for brining that. Now it makes sense to me. Actually custom unmarshall works fine when collector is running, but it isn't called in operatortest. 
It seems that operatortest calls unmarshall directly on the struct [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bd43e7355cea53f08a35256dab812d9c0076a7cf/pkg/stanza/operator/operatortest/confmap.go#L80)

So yeah, the problem with top-level config was that operatortest failed, because it didn't call custom unmarshall method.

To move this PR forward I can bring back top-level config and remove tests that are using operatortest for now. Not sure how to fix the issue with operatortest right now, probably it is better to do in the separate PR

djaglowski replied:
Ah, I see. That's unfortunate. We can revisit after we have a solution along the lines of https://github.com/open-telemetry/opentelemetry-collector/pull/6236

mar4uk committed:
[receivers/promtail] add comments

jpkrohling commented on .github/CODEOWNERS:
```diff
@@ -179,6 +179,7 @@ receiver/otlpjsonfilereceiver/                       @open-telemetry/collector-c
 receiver/postgresqlreceiver/                         @open-telemetry/collector-contrib-approvers @djaglowski
 receiver/prometheusexecreceiver/                     @open-telemetry/collector-contrib-approvers @dmitryax
 receiver/prometheusreceiver/                         @open-telemetry/collector-contrib-approvers @Aneurysm9 @dashpole
+receiver/promtailreceiver/                           @open-telemetry/collector-contrib-approvers @mar4uk
```
Please, add @kovrus and me as well. Perhaps even @gouthamve should be included :-)

mar4uk replied:
fixed

jpkrohling commented on .github/workflows/build-and-test.yml:
```diff
@@ -75,6 +75,9 @@ jobs:
     runs-on: ubuntu-latest
     needs: [setup-environment]
     steps:
+      - run: sudo apt-get -qq update
```
Could you add a comment stating that this is required because of the promtail receiver component? I read your comment that this is required for compiling this component. In that case, would the whole build fail if this component isn't present?

mar4uk replied:
By promtail receiver component isn't present you mean the repo doesn't have it in the receiver directory as well as all its mentions in other places like `internal/components.go`, `internal/components_test.go`, `versions.yaml`. Basically how it is now in the main branch, right?

If so the build will pass with or without installing libsystemd-dev.

jpkrohling replied:
We could isolate the code that depends on libsystemd-dev on a specific source, which is used only when a specific build flag is passed. The makefile could then check whether this lib exists and set the appropriate build flag. For our CI, we should always enable this, but I wouldn't require people on Mac or Windows systems to have this library to build contrib locally.

This brings another question: would the promtail receiver work on Windows and Darwin builds?

mar4uk replied:
Promtail receiver will work on Windows and Darwin builds because journal target is built only for linux and when cgo is enabled: https://github.com/grafana/loki/blob/488e64e3ff1bbce70d856c3daed150b061c5614d/clients/pkg/promtail/targets/journal/journaltarget.go#L1-L2

journal target will be ignored by non-linux OS and the collector build will be successful

> We could isolate the code that depends on libsystemd-dev on a specific source, which is used only when a specific build flag is passed. The makefile could then check whether this lib exists and set the appropriate build flag.

[promtail depends on libsystemd-dev](https://github.com/grafana/loki/blob/488e64e3ff1bbce70d856c3daed150b061c5614d/clients/pkg/promtail/targets/journal/journaltarget.go#L13), and promtail receiver depends on promtail. Could you please reveal the idea in more detail? What do you mean by a specific flag? Do you mean to have a custom build flag like `linux-with-libsystemd-dev`? To make it work I will need to add comment `//go:build linux-with-libsystemd-dev` to all promtail receiver files, right? And to run build we will pass flag `go build -tags linux-with-libsystemd-dev` That will work. Or if Loki team is agree we can add specfic flag `linux-with-libsystemd-dev` to https://github.com/grafana/loki/blob/main/clients/pkg/promtail/targets/journal/journaltarget.go to isolate that code and exclude it from compilation in case linux without libsystemd-dev. It will allow exclude exactly journald part from promtail without excluding the whole promtail receiver

mar4uk replied:
maybe more clear approach would be:
- check OS in Makefile
- if OS == linux then install `libsystemd-dev` package 
@jpkrohling what do you think?

mar4uk replied:
my last comment could be ignored

jpkrohling commented on cmd/configschema/go.mod:
```diff
@@ -1007,3 +1040,11 @@ exclude github.com/docker/distribution v2.8.0+incompatible
+
+replace github.com/hpcloud/tail => github.com/grafana/tail v0.0.0-20220426200921-98e8eb28ea4c
+
+// using fork for support LogpullReceived and LogpullFields for loki - see https://github.com/cloudflare/cloudflare-go/pull/743
```
Looks like the Loki team does not need this PR anymore. Can you check with them what the status is on this? The replace statement is a good temporary solution, but if that PR isn't being merged, this isn't temporary anymore.

jpkrohling commented on cmd/configschema/go.mod:
```diff
@@ -1007,3 +1040,11 @@ exclude github.com/docker/distribution v2.8.0+incompatible
+replace github.com/cloudflare/cloudflare-go => github.com/cyriltovena/cloudflare-go v0.27.1-0.20211118103540-ff77400bcb93
+
+// using fork for custom dialer support - see https://github.com/bradfitz/gomemcache/pull/86
+replace github.com/bradfitz/gomemcache => github.com/grafana/gomemcache v0.0.0-20220812141859-1e3ae89e91a7
```
Is this library abandoned? No commits since January, no commits in 2021 or 2020, and I see that you haven't received any answers on your PR yet :-/

mar4uk replied:
yeah, the library looks dead :(

mar4uk replied:
I received answers and did fixes. Waiting for the second round of review. Will keep an eye on this

jpkrohling commented on receiver/promtailreceiver/internal/logger.go:
```diff
@@ -0,0 +1,151 @@
+	return levelVal, true
+}
+
+//revive:disable:error-return
```
Do we need this?

mar4uk replied:
without this comment, golangci-lint throws an error: 
```
error-return: error should be the last type when returning multiple items (revive)
```

This code isn't mine actually, I copied it from prometheusreceiver/internal/logger.go and didn't refactor it.
I was thinking to move this code somewhere outside the components and import to promtail and prometheus receivers, but not sure if it is a good approach. 
If I move the logger somewhere (to internal directory maybe?) it will be there even if promtail and prometheus components are not present in the custom setup of collector contrib, so it will be an extra code user have to have that is not used anywhere

jpkrohling commented on receiver/promtailreceiver/promtail.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+func init() {
+	operator.Register(operatorType, func() operator.Builder { return NewConfig() })
```
What happens when this is executed twice? If it would fail, then wrap this on a `sync.Once`.

mar4uk replied:
init() function is called only once, so `operator.Register` will be called only once as well

mar4uk replied:
I checked what `operator.Register` does. It basically fills the `map[string]func() Builder`
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ff0a5123be74c6e2a59f3631c640e2753261d358/pkg/stanza/operator/registry.go#L35

If `operator.Register` executes twice, the builder function in the map will be overridden. No errors will occur.
I would keep `init()` function as it is because it looks the same way in all stanza receivers. Having `sync.Once` only in one stanza receiver (promtail receiver) will raise questions in the future. 
From my understanding adding `sync.Once` to `init()` function in all stanza receivers is not necessary because executing `operator.Register` twice will not lead to failure.

@djaglowski please correct me if I'm wrong. If we still need `sync.Once` in stanza receivers `init()` function I would do it in separate PR, because this change should be general for all stanza receivers I think

djaglowski replied:
@mar4uk, your assessment appears correct to me. You're using the same registration pattern in the `pkg/stanza` operators.

jpkrohling reviewed (commented):
Thanks for this PR! I had a few questions but looks good.

mar4uk committed:
[receivers/promtail] update CODEOWNERS

mar4uk committed:
[receivers/promtail] receiver/promtailreceiver go.mod

mx-psi reviewed (commented):
Can this component work without cgo? It looks like some functionality needs cgo https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14632#discussion_r1001663634, which [we don't support.](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#using-cgo)

mar4uk commented:
@mx-psi yeah, that's right, I recently found out that recommendation regarding cgo in collector.
Currently, I'm thinking about how to exclude code that depends on libsystemd-dev library. 
There are 2 ways to solve this:
- add custom build flag to promtail itself, so it will not include journald support if the flag is not passed
- add custom build flag to promtail receiver, so promtail receiver will not be part of binary by default.

If a user still wants journald support from promtail, then we will provide a clear readme guide on how to enable it (pass a custom flag, enable cgo, run on linux, install libsystemd-dev)

mx-psi commented:
Either one sounds like a reasonable solution to me :+1:  We now have [support for passing build tags on the builder](https://github.com/open-telemetry/opentelemetry-collector/blob/3262fd8f8bf990597bc5224c3b6f55988568808d/cmd/builder/internal/builder/config.go#L58) so the experience wouldn't be that bad

mar4uk committed:
[receivers/promtail] remove installing libsystemd-dev from build-and-test workflow

mar4uk committed:
[receivers/promtail] added promtail receiver to component dropdown

mar4uk commented:
Added go build tag `promtail_journal_enabled` to promtail itself https://github.com/grafana/loki/pull/7587.
It allows excluding journal support with C modules. Waiting for code review by loki team

mar4uk committed:
Promtail: update go.mod, README

mar4uk committed:
Promtail: fix failed build

mar4uk commented:
@jpkrohling @mx-psi could you please have another look?
I fixed the problem with promtail requiring CGO enabled and systemd libraries for journal support by introducing go build flag tag `promtail_journal_enabled`.
I [introduced this tag in promtail](https://github.com/grafana/loki/pull/7587) itself, so by default the binary will not contain C modules without this tag passed to go build

For collector, it means that by default promtail journal support is disabled, no need to install systemd headers for Linux/CentOS users, and enable CGO if the user doesn't need promtail journal support

If users need promtail journal support they will need to enable CGO, install systemd headers and pass build tag `promtail_journal_enabled`. I added this information to promtail receiver README

P.S 
build-and-test / integration-tests (pull_request) fails on mongodbreceiver... My changes shouldn't affect that receiver, maybe the test is flaky

mx-psi commented on receiver/syslogreceiver/go.mod:
```diff
@@ -37,11 +37,11 @@ require (
 	go.uber.org/multierr v1.8.0 // indirect
 	go.uber.org/zap v1.23.0 // indirect
-	golang.org/x/net v0.0.0-20220722155237-a158d28d115b // indirect
+	golang.org/x/net v0.0.0-20220812174116-3211cb980234 // indirect
```
How is this related to the promtail receiver?

mar4uk replied:
this is how `make gotidy` works. I added loki dependencies, and they need higher versions of `golang.org/x/net` and `google.golang.org/genproto`. That's why components that have dependency on `golang.org/x/net` and `google.golang.org/genproto` were updated

mx-psi replied:
I don't think that's the case, for example `make gotidy` did not bump the `golang.org/x/net` dependency on `internal/kubelet`: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c10532dc4e76de3a703fb485afcffae663dc7b8c/internal/kubelet/go.mod#L40 and the older version remains in your branch if I didn't misread anything.

I think what happened is that for some reason the `pkg/stanza` go.mod got modified (maybe you did it manually or maybe something else triggered it), which caused these log-related receivers to have their go.mod modified. If you don't have a good idea as to why `pkg/stanza` got modified, I would suggest to try and revert the changes, they are not strictly related to the promtail receiver and I wouldn't want to run the risk of introducing unintended changes.

mx-psi replied:
Awesome :)

mx-psi commented on receiver/solacereceiver/model/v1/receive_v1.pb.go:
```diff
@@ -223,13 +223,12 @@ func (SpanData_TransactionEvent_Initiator) EnumDescriptor() ([]byte, []int) {
 // nanoseconds, since midnight, Jan. 1, 1970 UTC.
 //
 // Notes on the field numbers used:
-// - Field numbers 1-15 are used for attributes that are expected to be present
```
ditto, is this intentional?

mar4uk replied:
I ran `make golint` locally to make sure it will pass in CI. This command fails locally with error (it fails locally now on main branch as well):
```
ERRO Running error: 1 error occurred:
	* can't run linter goanalysis_metalinter: gofmt: can't extract issues from gofmt diff output "--- /Users/mar4uk/go/mar4uk/opentelemetry-collector-contrib/receiver/solacereceiver/model/v1/receive_v1.pb.go.orig\t2022-11-14 08:46:32\n+++ /Users/mar4uk/go/mar4uk/opentelemetry-collector-contrib/receiver/solacereceiver/model/v1/receive_v1.pb.go\t2022-11-14 08:46:32\n@@ -223,13 +223,12 @@\n // nanoseconds, since midnight, Jan. 1, 1970 UTC.\n //\n // Notes on the field numbers used:\n-// - Field numbers 1-15 are used for attributes that are expected to be present\n-//   on the wire with every single message not containing an error_description.\n-//   Special priority is given to fields that can be repeated.\n-// - Field numbers 16+ are used for other attributes.\n+//   - Field numbers 1-15 are used for attributes that are expected to be present\n+//     on the wire with every single message not containing an error_description.\n+//     Special priority is given to fields that can be repeated.\n+//   - Field numbers 16+ are used for other attributes.\n //\n // Next available field ID: 39\n-//\n type SpanData struct {\n \tstate         protoimpl.MessageState\n \tsizeCache     protoimpl.SizeCache\n": can't parse patch: parsing time "2022-11-14 08:46:32" as "2006-01-02 15:04:05 -0700": cannot parse "" as "-0700"
```
That's why I ran `make gofmt` locally and commited changes. Interesting that this error doesn't occur in CI...

mar4uk replied:
reverted

mar4uk committed:
Promtail: update go.mod

mar4uk committed:
Promtail: revert pkg/winperfcounters/internal/third_party/telegraf/win_perf_counters/performance_query_test.go

mx-psi commented on go.mod:
```diff
@@ -1036,6 +1073,19 @@ retract v0.37.0 // Contains dependencies on v0.36.0 components, which should hav
+// using fork for custom dialer support - see https://github.com/bradfitz/gomemcache/pull/86
+replace github.com/bradfitz/gomemcache => github.com/grafana/gomemcache v0.0.0-20220812141859-1e3ae89e91a7
+
+// fork containing a line-buffered logger which should improve logging performance.
+// TODO: submit PR to upstream and remove this
```
I am worried about the large number of `replace` directives used here. `replace` directives hinder usability of the Opentelemetry Collector Builder since the user needs to understand what replace directives are needed for including a given component. They also make life harder for non-official distros, which run into basically the same problem.

My suggestion here would be to instead require these forks directly. This may need changes on the dependencies you use, but it's the only way to avoid this for builder users

mar4uk replied:
I made commit messages more clear, so it is more visible for users which `replace` is needed for loki.
I went through all of them and tried to get rid of some of them as much as possible, sent PR upstream, and discussed them with loki team. 
I understand that replaces are not convenient, but it is difficult to get rid of them right now. 
Requiring forks instead of upstream is not a clean solution for loki, since it will lead to a lot of changes in loki repo itself and in forks as well. Also, it will take more effort to switch back to the upstream in the future and will discourage from creating PRs to the upstream.

I would leave these replaces. And since I'm one of the owners of components that depend on loki, I will keep an eye on these replace and try to remove them

mx-psi replied:
Since the problems are scoped to builder users that want to use the `promtail` receiver, I will leave this up to @jpkrohling

jpkrohling replied:
I suggest the following compromise: we'll commit to reaching out to the Loki team and expose this problem, working with them on a solution that suits us all. In the meantime, I'd suggest having the minimal set of replaces required to get this PR into a working state.

mx-psi replied:
Okay, sounds reasonable to me. I think we also need guidelines on replace directives more generally, I will look into that from my side :) Please post the tracking issue here once created so that future readers can follow the conversation

jpkrohling replied:
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16453

mar4uk committed:
Promtail: revert receive_v1.pb.go, revert stanza go.mod changes

mar4uk committed:
update promtailreceiver go.mod

mar4uk commented on go.mod:
```diff
@@ -1036,6 +1073,19 @@ retract v0.37.0 // Contains dependencies on v0.36.0 components, which should hav
 // see https://github.com/distribution/distribution/issues/3590
 exclude github.com/docker/distribution v2.8.0+incompatible
 
-// some dependencies attempt to bring something like v1.8.2-0.20220303173753-edfe657b5405, which is older than v0.38.0
-// at the time of this inclusion, v0.38.0 was the latest version available (also tagged as v2.38.0)
-replace github.com/prometheus/prometheus => github.com/prometheus/prometheus v0.38.0
+// some dependencies attempt to bring something like v1.8.2-0.20220303173753-edfe657b5405, which is older than v0.39.1
+// at the time of this inclusion, v0.39.1 was the latest version available (also tagged as v2.39.1)
+replace github.com/prometheus/prometheus => github.com/prometheus/prometheus v0.39.1
```
created PR to loki to get rid of this replace https://github.com/grafana/loki/pull/7697
As soon as it is merged I will remove replace from the collector. But I think it is ok to do it in the separate PR, since this replacement is already present in the main branch

mx-psi replied:
Sure, this shouldn't block this PR,  let's just create an issue to track this and we can mark it as solved

mar4uk replied:
https://github.com/grafana/loki/pull/7697 was merged, so I removed prometheus replace directive from go.mod in this PR instead of creating the new one

mar4uk committed:
remove replace prometheus directive from go.mod

jpkrohling commented on receiver/promtailreceiver/README.md:
````diff
@@ -0,0 +1,92 @@
+
+With Journal support on CentOS, run with the following commands:
+```shell
+sudo apt install -y libsystemd-dev
````
Sorry for catching this only now, but aren't the commands inverted? CentOS would use `yum` and Ubuntu uses `apt`.

mar4uk replied:
oops, thank you! you are right

jpkrohling reviewed (approved):
LGTM. There are a few comments, but I believe this can be merged as is and a follow-up PR can address the remaining points.

jpkrohling commented:
@djaglowski, I believe this is ready, but I will merge it once you approve. Thank you very much for your review!

djaglowski reviewed (commented):
Generally LGTM, with a couple nits & outstanding comments from others.

jpkrohling commented on receiver/promtailreceiver/config.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// Build will build a promtail input operator from the supplied configuration
+func (c Config) Build(logger *zap.SugaredLogger) (operator.Operator, error) {
```
Oh, now it makes sense. Thanks for the clarification.

mar4uk committed:
fix README, promtail go.mod cleaning

mar4uk commented on receiver/promtailreceiver/go.mod:
```diff
@@ -0,0 +1,253 @@
+
+go 1.18
+
+replace github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza => ../../pkg/stanza
```
done

mar4uk commented on receiver/promtailreceiver/go.mod:
```diff
@@ -0,0 +1,253 @@
+	go.opentelemetry.io/collector v0.65.0
+)
+
+require (
```
fixed

mar4uk commented on receiver/promtailreceiver/internal/logger.go:
```diff
@@ -0,0 +1,151 @@
+	// need to skip two levels in order to get the correct caller
+	// one for this method, the other for log
+	logger = logger.WithOptions(zap.AddCallerSkip(2))
+	return &zapToGokitLogAdapter{l: logger.Sugar()}
```
agreed that a sugared logger should be used because it is required by Builder interface

mar4uk commented on receiver/promtailreceiver/README.md:
```diff
@@ -0,0 +1,92 @@
+
+[sc]: https://grafana.com/docs/loki/latest/clients/promtail/configuration/#scrape_configs
+
+[beta]: https://github.com/open-telemetry/opentelemetry-collector#beta
```
thank you! fixed

mar4uk commented:
@djaglowski I fixed all outstanding comments. Could you please take a look and if everything is ok, approve it?

mx-psi commented on receiver/promtailreceiver/promtail.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// PromtailInput implements stanza.Operator interface
+type PromtailInput struct {
```
Does this need to be public?

mar4uk replied:
no, it doesn't, made it private

mx-psi commented on receiver/promtailreceiver/factory.go:
```diff
@@ -0,0 +1,63 @@
+
+// ReceiverType implements stanza.LogReceiverType
+// to create a promtail receiver
+type ReceiverType struct{}
```
Does this type need to be public?

mar4uk replied:
no, it doesn't, made it private

mx-psi commented on receiver/promtailreceiver/config.go:
```diff
@@ -0,0 +1,117 @@
+	adapter.BaseConfig `mapstructure:",squash" yaml:",inline"`
+}
+
+// NewConfig creates a new input config with default values
+func NewConfig() *Config {
+	return NewConfigWithID(operatorType)
+}
+
+// NewConfigWithID creates a new input config with default values
+func NewConfigWithID(operatorID string) *Config {
+	return &Config{
+		InputConfig: helper.NewInputConfig(operatorID, operatorType),
+	}
+}
```
Do these functions need to be public?

mar4uk replied:
`NewConfigWithID` should be public because it is called in `internal/components/receivers_test.go` https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/dde66116c4973842f19c3675e292f23d84966dfc/internal/components/receivers_test.go#L294
Made private other methods and struct you mentioned above. Thank you!

mx-psi replied:
Okay, let's leave that one public then

mx-psi reviewed (commented):
Just a few comments on whether certain types/functions need to be part of the public API, otherwise I have no other comments :)

mar4uk committed:
make some methods and struct private

mx-psi commented:
@mar4uk can you fix the merge conflicts? I think we are ready to merge after that

mar4uk committed:
update promtailreceiver go.mod

mar4uk committed:
update promtail receiver factory to not use deprecated type

mar4uk committed:
update go.sum and cmd/configschema/go.sum

jpkrohling reviewed (approved):
LGTM, thank you @mar4uk!
